### ISSUE TITLE
Allow for optional custom installer switches

### ISSUE THREAD

**Issue by Trenly** [microsoft/winget-cli#1783]
### Description of the new feature / enhancement

In the manifest schema, there is a field for custom installer switches, which are always applied. However, packages may support other custom switches that users may want to use without having to use the override parameter. If there was an ability to add these as optional switches and to show them in `winget show`, it could make for a better overall experience

### Proposed technical implementation details

Thought 1 - Change `InstallerSwitches > Custom` to an array type of `CustomSwitch` objects
```yaml
InstallerSwitches:
    Silent: "/quiet"
    Custom:
    - Switch: NoRestart
      Value: "/suppressReboot"
      Default: true
    - Switch: NoDesktopIcon
      Value: "/desktopicon=0"
      Default: false
```
User would run `winget install example.package --silent --NoDesktopIcon` and the package would be installed using `example.package.exe /quiet /suppressReboot /desktopicon=0`

-----
Thought 2 - entirely new manifest field
```yaml
InstallerSwitches:
    Optional:
        NoDesktopIcon: "/desktopicon=0"
        UseAppdataFolder: "/appdata=1"
```
User could run `winget install example.package --UseAppdataFolder`

**Comment by jedieaston** [comment]
I really like this idea, as having to know the magic incantation to install VS with the right workloads is really annoying. 

It may present a problem for automated validation, because it would have to test all of those arguments. Could possibly slow it down a ton.

**Comment by Trenly** [comment]
> It may present a problem for automated validation, because it would have to test all of those arguments. Could possibly slow it down a ton.
>
I do agree that it could slow it down, likely significantly. However, given that validations can run in parallel, as seen by each PR being validated individually rather than part of a queue (so long as the resource pool has a free resource), it may be possible to run all of the validations in parallel async for a single PR. 

The other factors that could weigh in here are the number of packages that could, or would, receive the optional switches. As you mentioned, it could be very useful for some packages like VisualStudio, and I expect that they would be used there. However, there are (probably) a lot of installers which support location switches which do not have them, and I would expect a similar behavior here. That behavior being that the packages start without them, but as the repository matures and tools like YamlCreate and Wingetcreate mature also, the fields are much more likely to become populated but I would still expect it to be a minority percentage of packages that would take advantage of this.

**Comment by gnpaone** [comment]
> ### Description of the new feature / enhancement
> In the manifest schema, there is a field for custom installer switches, which are always applied. However, packages may support other custom switches that users may want to use without having to use the override parameter. If there was an ability to add these as optional switches and to show them in `winget show`, it could make for a better overall experience
> 
> ### Proposed technical implementation details
> Thought 1 - Change `InstallerSwitches > Custom` to an array type of `CustomSwitch` objects
> 
> ```yaml
> InstallerSwitches:
>     Silent: "/quiet"
>     Custom:
>     - Switch: NoRestart
>       Value: "/suppressReboot"
>       Default: true
>     - Switch: NoDesktopIcon
>       Value: "/desktopicon=0"
>       Default: false
> ```
> 
> User would run `winget install example.package --silent --NoDesktopIcon` and the package would be installed using `example.package.exe /quiet /suppressReboot /desktopicon=0`
> 
> Thought 2 - entirely new manifest field
> 
> ```yaml
> InstallerSwitches:
>     Optional:
>         NoDesktopIcon: "/desktopicon=0"
>         UseAppdataFolder: "/appdata=1"
> ```
> 
> User could run `winget install example.package --UseAppdataFolder`
> 
> Edit: Thought 3 - Allow for them to be under Installer Switches
> 
> ```yaml
> InstallerSwitches:
>     Silent: “/quiet”
>     Custom: “/norestart”
>     NoDesktopIcon: "/desktopicon=0"
>     UseAppdataFolder: "/appdata=1"
> ```

Is there any progress on this? Thought 1 is great

**Comment by Trenly** [comment]
[Policy] Area-Manifest

**Comment by s-crypt** [comment]
Number one would be great for software that is mass deployed. Especially since lots of the deployed software takes in the license key as an installer switch or have their own weird custom flags.
Something like 
```yaml
InstallerSwitches:
    Silent: "/quiet"
    Custom:
    - Switch: Port
      Value: "/portable "              # Note space after switch. some need spaces some don't, so have manifest creator determine what is necessary
      Default: true                    # Default means switch automatically applied
    - Switch: Licensekey 
      Value: "/LICENSE="
                                       # Lack of default means switch is not automatically applied


```
`winget install example.msi --LicenseKey 123456789abc`
would result in 
`msiexec /i example.msi /portable /LICENSE=123456789abc`
as the order of the switches would be preserved

Enterprise software sucks. However, this would make it suck just a little less for those who have to deploy the stuff.

**Comment by marcinsmialek** [comment]
That makes a lot of sense.
However the optional non-switch parameters like key, may be addressed in another issue, which might be worth linking to.

**Comment by s-crypt** [comment]
Do you have a link to that issue? Since situations like `/LICENSE=` are parameters/switches, I think we could kill two birds with one stone here and just add it to the docs for both use cases.

**Comment by mdanish-kh** [comment]
If it's helpful to you, you can already do that with the `--custom` switch. `winget install Example.Package --custom "/LICENSE=<key>"` will append the switch at the end of the default switches.

I believe this issue is mainly for having support in the manifest for a better UX

**Comment by s-crypt** [comment]
Yes, the custom switches are very useful, but the documentation of having the possible switches as part of the manifest would be very helpful as they vary between programs.

I believe this would also help Microsoft/Sysadmins in Intune as there could be native support for deploying software with the switches already listed.
